### PR TITLE
Replace design screenshots with mShots in New Onboarding

### LINKED ISSUE #48867: New Onboarding: add translated design thumbnails
At translated version of the New onboarding flow (https://wordpress.com/new/design/es), design thumbnails are in English.

Let's get them translated, so that the first impression is better for non-English speakers.

Screenshot 2021-01-13 at 15 42 10

In the preview translations work just fine:

Screenshot 2021-01-13 at 15 42 03

Perhaps for starters we just generate static screenshots for all the locales we support? I know it's a lot of images ...but so what? 😅

### PR DESCRIPTION
**Pull request by andrewserong**
#### Changes proposed in this Pull Request

* Generate an mShots screenshot for the design preview that passes in a locale to the `template/demo` endpoint
* Add an `MshotsImage` component that can deal with hiding the mShots default loading gif, and continue to request the image until it loads the final screenshot. The user should only see a pulsing rectangle until the image finally loads

#### Outstanding issues and todos

* [ ] The vertically oriented screenshots aren't _quite_ how we'd like them, some of the viewport height based elements appear stretched vertically a little bit more than desired. This may or may not be a deal-breaker.
* [ ] Currently the image file size is a bit too much greater than the existing webp approach. Separately to this PR we're proposing adding in a quality param for mShots to support smaller file sizes.
* [ ] The logic to keep the image hidden until _really_ loaded is a bit hacky and relies on a hard-coded URL. Is this safe enough, and does it hold up to testing when generating new screenshots? Is there a reliable way to test this, or write a test for it?
* [ ] Are we caching correctly? If a user goes back and forward between steps, does the browser re-download the images, or does it correctly cache them? We should ensure that the images aren't being re-downloaded multiple times.

#### Screenshot

mshots-in-design-step-small

#### Testing instructions

* Go to Calypso live URL below and append the mshots flag to the end of the URL, e.g. add `?flags=gutenboarding/mshot-preview` or build locally and run `calypso.localhost:3000/new/ko?flags=gutenboarding/mshot-preview`.
* Either set your user language, or add a locale to the URL and go to Gutenboarding, e.g. `/new/ko?flags=gutenboarding/mshot-preview` or `/new/de?flags=gutenboarding/mshot-preview`
* At the design step, the previews should be translated
* Test this in a number of different languages
* Test this with limited Network speed (e.g. Fast 3G or Slow 3G in the Network tab in dev tools) — this will demonstrate that the file sizes for the images are currently too big.

Related to #37665 
Fixes #48867

### CODE REVIEW

**Comment by andrewserong** [comment]
It seems to be working okay so far, specifying a large viewport height max, however designs that use viewport height in their layouts render strangely (or, well, accurately to their CSS and the viewport they're dealing with, but not how we want them to):

image

**Comment by simison** [comment]
Yeah, Reynolds (and perhaps other designs?) need a bit extra CSS help somehow: https://github.com/Automattic/wp-calypso/blob/1227e50feb9c6b0ea482717eefb05b28257ded2f/bin/generate-gutenboarding-design-thumbnails.js#L59-L64

**Comment by simison** [comment]
In case this is useful: https://github.com/Automattic/wp-calypso/blob/4537033c6794449ce0a10b69d72daf5fa658d553/client/lib/mshots/index.js

**Comment by andrewserong** [comment]
Thanks @simison that CSS does the trick. I've worked up a diff for adding the override to the `template/demo` endpoint, which should do the trick for mshots. (D45189-code)

File: client/landing/gutenboarding/available-designs.ts
Comment (simison) [inline client/landing/gutenboarding/available-designs.ts]: We could also add cache-busting timestamp coming from templates API?

File: client/landing/gutenboarding/available-designs.ts
Comment (simison) [inline client/landing/gutenboarding/available-designs.ts]: Might be better to have `700` in constant so that nobody misses updating it between `previewUrl` and `mshotsRequest`?

**Edit:** Oh, I just realized this was height _and_ width, not _height_ and _height_ 😁

File: client/landing/gutenboarding/available-designs.ts
Comment (simison) [inline client/landing/gutenboarding/available-designs.ts]: Sadly this is capped at `1200` currently:

https://github.com/Automattic/mShots/blob/100faf8c8c1148fd8584713999ab944e6fcf24a3/public_html/class-mshots.php#L15

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (simison) [inline client/landing/gutenboarding/components/mshots-image/index.tsx]: if `status` is something else (e.g. an error 500), we could just keep looping otherwise for a while and then give up by disabling the preview thumbnail?

Not very high priority or blocking this PR though.

File: client/landing/gutenboarding/onboarding-block/design-selector/index.tsx
Comment (simison) [inline client/landing/gutenboarding/onboarding-block/design-selector/index.tsx]: I think for initial roll-out we should keep `isEnabled( 'gutenboarding/mshot-preview' )` so that we can quickly revert to static method if there are problems.

File: client/landing/gutenboarding/available-designs.ts
Comment (deBhal) [reply client/landing/gutenboarding/available-designs.ts]: @simison: Seems to work ok?
https://s.wordpress.com/mshots/v1/https%3A%2F%2Fgithub.com%2FAutomattic%2Fwp-calypso%2Fpull%2F43428?vph=3072
vs
https://s.wordpress.com/mshots/v1/https%3A%2F%2Fgithub.com%2FAutomattic%2Fwp-calypso%2Fpull%2F43428?vph=1200

Comment (andrewserong) [reply client/landing/gutenboarding/available-designs.ts]: Yep, looks like the version we have running supports greater max width/height (p3topS-Ia-p2#comment-2967)

**Comment by andrewserong** [comment]
I'm wrapping up for the day and will be AFK until next week, when I'm then on gardening duty, so mightn't have time to work on this one for a couple of weeks. Feel free to pick this up, if anyone would like to!

Here are some notes / things to investigate from where I've gotten up to. While working on it I found it easiest to add `"gutenboarding/mshot-preview": true` to [development.json](https://github.com/automattic/wp-calypso/blob/ab5147f328547439677e0ce3f5c3f5eadd3f169a/config/development.json#L71) for testing it locally.

Some to dos:

* [ ] The mShots endpoint supports CORS requests, however in Google Chrome, I notice that the first time it loads it succeeds, but if I browse forwards or backwards a step, then the network requests appear to load from cache, where the Origin header doesn't appear to be sent. So the response from the server doesn't have the `access-control-allow-origin: http://calypso.localhost:3000` header, causing the images to fail to load. This will need to be investigated / resolved for this particular implementation to work in production.
* [ ] mShots URLs that 404 (e.g. the site that it's taking a screenshot of doesn't exist) results in a 404 image (with 200 response code) instead of a 404 response, so we can't really check for a 404 in our code and handle it. The 404 image is:

mshots-404

* [ ] Update the way the `MshotsImage` component fetches from the server, how many times do we want to check for a 307 before bailing?
* [ ] In the event of responses in the ranges 400 and 500, how do we want to handle what happens with the image?
* [ ] In testing locally, getting these images from mShots, even when they've already been generated, appears to be a bit slower to load than images stored in `static`. We should work out how to either a) make sure the images are loaded / cached earlier in the process, like we're currently doing for the static images when Gutenboarding loads, or assess whether we should still go with static images, but use mShots to generate them.

Just some thoughts, feel free to change / go in a different direction if you happen to pick up this PR.

Even though there's a fair bit to getting this working smoothly, the vertical screenshots are looking pretty good, and I'm happy so far with how the loading / onload state is looking:

mshots-design-previews-small

File: client/landing/gutenboarding/available-designs.ts
Comment (simison) [reply client/landing/gutenboarding/available-designs.ts]: Oh, that's wonderful! I had missed an update in that P2. (_How could I!_)

**Comment by simison** [comment]
> mShots URLs that 404 (e.g. the site that it's taking a screenshot of doesn't exist) results in a 404 image (with 200 response code) instead of a 404 response, so we can't really check for a 404 in our code and handle it. The 404 image is:

We could add a custom header in response at mshots?

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (simison) [inline client/landing/gutenboarding/components/mshots-image/index.tsx]: An alternative, just for your consideration:

```js
const img = new Image();
img.addEventListener('load', (event) => {
 // event
}, false);
img.src = mShotsEndpointUrl;
```

Might help with cors stuff.

Comment (andrewserong) [reply client/landing/gutenboarding/components/mshots-image/index.tsx]: Oh, that's a nice idea! Unfortunately, I think we might still need to use the fetch approach if we want to hide the fallback `/default` route that renders [the default loading gif](https://s0.wp.com/mshots/v1/default), by handling the 307 redirect case.

It just occurred to me that I haven't looked into the [link prefetch](https://github.com/Automattic/wp-calypso/blob/f392fd8fbdb062b79034b6c2650b5d1bdb687276/client/landing/gutenboarding/available-designs.ts#L62) that we're doing in `prefetchDesignThumbs` — my hunch is that we need to add [the `crossorigin` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content#Cross-origin_fetches), and this might explain why the CORS requests appear to be failing some of the time, but not when I directly reload the Design step. Just a hunch, though! I'll take a quick look early next week and see if it fixes it, if no-one beats me to it!

**Comment by deBhal** [comment]
> however designs that use viewport height in their layouts render strangely

Yeah, it's looking great overall but there are a few little oddities. Different things in different previews, so I think in addition to the css fixes we'll want to pass size/viewport parameters through from `available-designs-config.json` to be able to make them all appropriately attractive.

**Comment by deBhal** [comment]
@simison @andrewserong : I was looking at Rockfield, and the viewport_height param on the template demo endpoint helps, but the images still end up the wrong size and looking pretty bad.

I had a play with mshots, and I think this might be a better solution than special-casing the css, but I wanted to make sure I wasn't just geeking out. Do you think this approach is worth completing? : https://github.com/Automattic/mShots/pull/19

**Comment by andrewserong** [comment]
@deBhal nice, thanks for exploring this! It'd be worth commenting about it on p3topS-Ia-p2 as well, since there's some earlier related discussion there.

Personally I think that's a good direction to go down, considering that we'd like to be able to support a wider range of designs than we currently have, and who knows what other kinds of workarounds we might need to add for special-casing CSS in the future. If mShots can behave a little bit more like a "real" 16:9 browser that's scrolled down the page, then it might save us time further down the track to handle it in mShots. That said, I have very little knowledge of how mShots works, and it's used in a lot of other places, too, so we'd need to be careful that it doesn't conflict with how any other services are using it.

The other param I kinda wish we could support is a custom `quality` param. Right now it's locked to 90, but to reduce file-size when we're displaying a whole bunch of them, it'd be nice if we could pass in, say, 70 or something (or have a controlled set of low, mid, high quality settings?)

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (simison) [inline client/landing/gutenboarding/components/mshots-image/index.tsx]: I know this might've been just temporary due lack of designs but let's use pulsating gray box instead of wall of spinners. 😁

Comment (andrewserong) [reply client/landing/gutenboarding/components/mshots-image/index.tsx]: Yep! It was just a placeholder to use in place of real designs, and something that looked better than the string “loading...” 😄

**Comment by ramonjd** [comment]
@deBhal Could you update the description and test plan so folks can start testing? Thanks!

**Comment by andrewserong** [comment]
This PR isn't _really_ quite ready for review, but I've rebased it and cleaned it up a bit so that we can test it with the translated previews, so I've switched the PR over from Draft. I think the file sizes of the images are too big for this to be ready for production (that feature is proposed in https://github.com/Automattic/mShots/pull/23, and likely needs some additional work). But at least we can get a couple more pairs of eyes on this PR and get a feel for how it's looking so far.

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (andrewserong) [reply client/landing/gutenboarding/components/mshots-image/index.tsx]: Updated to be a set of pulsating gray boxes instead of spinners.

**Comment by simison** [comment]
Do we pre-load these images early in the flow, before user lands on design picker? Feels important for this to work smoothly.

**Comment by lsl** [comment]
Rebased, fixed a conflict or two and updated locale lookup ([`i18nLocale` was removed](https://github.com/Automattic/wp-calypso/commit/2c8ffc9a6b2a4225cbf9752f16827369b8df37ff#diff-d9fe05747d9a2705393b29af2553c0e4ed4ee536495953647f8b1e06b6404db3L80).)

@andrewserong This looks pretty good. Only ran into cors problems with s0.wp.com sandboxed. Unsure if its related or not to your [earlier fix](https://github.com/Automattic/wp-calypso/pull/43428/commits/1a6dc6c7fc53e83e5ca236c0f5b32a58551db15e) or [if that even matters anymore?](https://github.com/Automattic/wp-calypso/commit/20b6787762f6717c79ebe420dccec7269706c801#diff-9040d891229d796c6bcab72e8c55a02eede9675da8552c87c0e94b6fd80d426dL36)

> Do we pre-load these images early in the flow, before user lands on design picker?

@simison [we used to](https://github.com/Automattic/wp-calypso/commit/20b6787762f6717c79ebe420dccec7269706c801#diff-9040d891229d796c6bcab72e8c55a02eede9675da8552c87c0e94b6fd80d426dL36), but that got dropped, not clear why, maybe @yansern recalls why the line `React.useEffect( prefetchDesignThumbs, [] );` was removed?

Adding the line back doesn't seem to make a big difference in the loading of /new but does re-enable prefetching. (provided you remember to turn prefetching back on in about:config :crossed_fingers: )

**Comment by andrewserong** [comment]
Thanks for picking this one up again @lsl ! I _think_ the CORS issues I ran into earlier no longer apply because of the removal of the image pre-load.

When I was last working on it, I noticed some odd behaviour with the loading state, and I'm not sure if the placeholder url logic to reload the image (when we get `default.gif` instead of the expected screenshot) is working quite right.

I just gave this a quick test in the current calypso.live url, and when I requested the design page directly in German (`/new/design/de?flags=gutenboarding/mshot-preview`), I could see the mshots loading screen (which we're trying to hide), and it didn't automatically swap out the loader for the real image when it had finished generating the screenshot (clicking back and then forward appears to resolve it): 

image

I _almost_ feel like I want there to be an endpoint in front of mShots to give us a JSON response with a status like `{ status: 'generating_screenshot' }`, and when it's generated gives us a `{ status: 'success', data: { url: 'https://mshots-url...' } }` response, and we grab _that_ url to place directly into an `img` tag. But perhaps that's overkill?

**Comment by lsl** [comment]
I'd like it if mshots gave us the option to request the screenshot in real time and we just deal with the long timeout ourselves using a placeholder/swapout.

I don't think it makes sense for the same url used to request the image does a redirect *sometimes* to a loading gif. That behavior pretty hard to make use of, especially as the image doesn't then replace itself (because its just a gif)

**Comment by simison** [comment]
p3topS-Ia-p2 has a bit about "Blocks HTTP requests until the screenshots are ready instead of redirecting to a loading GIF." which would solve some of this.

It's a no-no to show those uggly loading screens to customers so that's one of the reasons why I was interested if we load these es early in the `/new` float as possible (in addition to being able to show them sooner when they land in design picker). I think the generated image stays in Mshots for like a day or so.

File: client/landing/gutenboarding/available-designs.ts
Comment (simison) [inline client/landing/gutenboarding/available-designs.ts:39]: If we exclude `site_title`, we'd more often have previews generated already instead of needing to generate for each customer separately.

That said, if it works, would be _really cool_ to include it in thumbnails. 😁

**Comment by ramonjd** [comment]
Could we perhaps just yield to the weaknesses of mshots and show landscape for now? Would that unblock this?

Context: pbxNRc-vn-p2#comment-1425

cc @simison @ollierozdarz

**Comment by simison** [comment]
If it moves us forward with Mshots, let's go landscape (even temporarily).

**Comment by lsl** [comment]
Debugged the `MshotsImage` component a bit, the logic looks ok but we run into some timing issues where one request might respond with the image being available so we swap the image out for the new one but then the `` loads up a default image anyway. Could be cache, could be server responding differently between requests.

Screenshot_2021-01-19 WordPress com(1)

We might be able to flip this around a bit but I'm thinking we may end up being forced fixing this on the mshots side.

**Comment by lsl** [comment]
Updated `<MShotImage>` to load the image using  directly instead of issuing the ajax request. We then update the src attribute repeatedly until we get the non-default image coming back. This usually completes for all images < 5 seconds but it depends on mshots and the interval we use. Prefetching in the aquire intent (title) step makes this a lot smoother but we also don't risk any cache miss like before if anything is ejected in between long (domain) steps or if mshots simply drops a cached image (seems to happen)

#### Prefetched:
prefetched

#### vs No Prefetch:
no-prefetch

I've included a delay attribute but It really just slows things down to make it seem like the first 4 images load faster (for the most part), this might have a benefit if we can delay initial render so the first 4 images (above fold) really do get a benefit from having a headstart.

3G testing wasn't too bad, but as expected it was relatively slow comparatively. 

Next steps: Remove the delay and deploy behind feature flag.

File: client/landing/gutenboarding/available-designs.ts
Comment (lsl) [reply client/landing/gutenboarding/available-designs.ts:39]: This is actually the theme title (maybe it wasn't previously?) I think I'll try using users title tomorrow, see how it goes, prefetching seems to work decently well at the moment.)

Comment (simison) [reply client/landing/gutenboarding/available-designs.ts:39]: Ahh, gotcha! Users title would be awesome of course — it'll just potentially generate a whole lot more mshots images once this is live to more people, so might be good to give systems heads up?

Also just slower since every image needs to be generated for each user, but good to test how that tradeoff looks like? My preference is on speed.

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (simison) [inline client/landing/gutenboarding/components/mshots-image/index.tsx]: One thing to try is put ridiculously low jpg quality, and pretty big physical image size.

When rendered in small enough canvas, it sometimes looks better and loads faster than better quality, but smaller physical size image. 😄 

Unless mshots is smart and returns webp for browsers that support it? 🤔

**Comment by simison** [comment]
> I'd like it if mshots gave us the option to request the screenshot in real time and we just deal with the long timeout ourselves using a placeholder/swapout.

My worry is just that the loading time would be long and people see loading squares, especially as we add more designs. If it's fast regardless, then ignore me. :-)

What if we load them early in `` but outside visible page?

**Comment by lsl** [comment]
Pulled together current todo's / suggestions into #49097 so we don't lose any thoughts after merging this PR.

**Comment by lsl** [comment]
Dropped the delay commit, rebased in hopes e2e tests are working better now :crossed_fingers: 

I'm seeing the CORS issue with prefetching now (in chrome.) Will address in the follow up, images still load fine so I suspect its just the way we're making the request.

> What if we load them early in  but outside visible page?
Might be a good fix for this.

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (deBhal) [inline client/landing/gutenboarding/components/mshots-image/index.tsx:32]: have you tried `requeue: "true",`?

Comment (lsl) [reply client/landing/gutenboarding/components/mshots-image/index.tsx:32]: No, but if that's it you may save me some time in the future :+1:

File: client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx
Comment (deBhal) [inline client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx]: Relatively minor, but can we put the `React.useEffect()` inside the config check?

File: client/landing/gutenboarding/onboarding-block/design-selector/index.tsx
Comment (deBhal) [inline client/landing/gutenboarding/onboarding-block/design-selector/index.tsx]: These curly braces should probably go.

File: client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx
Comment (lsl) [reply client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx]: Nope! Did that first but linter complains about hooks needing to be consistently run, its probably safe (config won't change mid run) but it was easier to just do what it says.

File: client/landing/gutenboarding/onboarding-block/design-selector/index.tsx
Comment (lsl) [reply client/landing/gutenboarding/onboarding-block/design-selector/index.tsx]: I think I'll just remove the comment completely, was moved from somewhere else but its probably no longer valid.

File: client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx
Comment (lsl) [reply client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx]: Decided to comment this out completely until CORS issues are resolved.

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (deBhal) [inline client/landing/gutenboarding/components/mshots-image/index.tsx]: This 1800 looks pretty weird here. Not a blocker for this PR, but it'd be nice if we could make this the same value as we pass to mshots. Maybe with an argument to `mshotsUrl()`?

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (deBhal) [inline client/landing/gutenboarding/components/mshots-image/index.tsx]: If I understand correctly, this is waiting a full second between the getting the response back from mshots and triggering the next one. I'm seeing time-to-first-bytes of 600ms or more on both the initial query and the `default.png` request (curiously, this includes when chrome reports the image is served from the disk cache - I guess the server must be taking this long to send back a timestamp? )

Anyway, I'm seeing up to two seconds between successive requests for the same image e.g. 5.35s, to 7.57s and then 9.15s and finally 10.74s to initiate the the final successful request.

We totally shouldn't DDOS mshots, but I think we can bring this down. I'd suggest we throttle to 1s rather than wait that long after we get a result.

On the other hand, I saw it get up to `count=3` and still get the `307` 9 seconds after page load, so this is clearly _not_ the bottleneck in the process (*cough* mshots *cough*), so if you agree let's put it into the followup list rather than this PR.

**Review by deBhal — APPROVED** [review]
I haven't looked into the e2e failures, but we discussed in chat that they're not related to this PR, so assuming that's the case, I say :shipit: .

There are a number of issues with this approach, but this PR is a clever & working solution to some of them.

My biggest concern is the 10 second delay between page load and starting the request that actually succeeds. It's hard to see mShots being a viable approach if it's going to be common for users to be waiting that long on rarer locales.

Again, though, this code is working impressively well, good job!

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (lsl) [reply client/landing/gutenboarding/components/mshots-image/index.tsx]: Follow up for this I think, it definitely needs some tweaking, but too low results in a sluggish browser as it tries to load 20 images x 3-5 times.

File: client/landing/gutenboarding/components/mshots-image/index.tsx
Comment (lsl) [reply client/landing/gutenboarding/components/mshots-image/index.tsx]: Done (well sort of https://github.com/Automattic/wp-calypso/pull/43428/commits/900d933bfcaa9698c2b3ac973f300bfc350391ac)

**Pull request merged**
